### ISSUE TITLE
how to assign trainable variable to a tensor to form another trainable variable

### ISSUE THREAD

**Issue by chencan15** [tensorflow/tensorflow#26819]
### System information

TensorFlow version : 1.12
Are you willing to contribute it : Yes

### Describe the feature and the current behavior/state.

This issue occurs in my nerual network. The data A stored in TFrecord file is with shape [batch_size,  A_size * A_size]. I want to use a fully connected layer to increase the dimension to get a bigger tensor B with shape [batch_size, B_size * B_size]. Then zero-padding B to get a bigger C with shape [batch_size, C_size, C_size]. 

> Note: B[i,:,:] is a block extracted from A[i,:,:][position_height_index * B_size:(position_height_index + 1) * B_size, position_width_index * B_size:(position_width_index + 1) * B_size]
 
Here is my code:
```python
input_size = A_size * A_size
output_size = B_size * B_size

 w = tf.Variable(tf.random_normal([input_size, output_size]))
 b = tf.Variable(0.0)

# A is a data tensor with shape [batch_size,  A_size * A_size]

 prediction = tf.matmul(A,w)+b

 C = tf.Variable(np.zeros([batch_size, C_size * C_size]), True, dtype = tf.float32)
 
for i in range(batch_size):
            
            # initialize temp
            temp = tf.Variable(np.zeros([C_size, C_size]), True, dtype=tf.float32)
            
            # assign reshaped prediction to temp with corresponding positions
            temp[position_height_index[i] * B_size:(position_height_index[i] + 1) * B_size,
            position_width_index[i] * B_size:(position_width_index[i] + 1) * 
            B_size].assign(tf.reshape( prediction[i], [B_size, B_size]))
            
           # assign reshaped temp to C
           C[i].assign(tf.reshape(temp, [1, C_size * C_size]))
     
  ```
### My Question: 
When i print C, all elements in C are still zero. It should be non-zero because i assign prediction to C, what's wrong. I assign trainable variable prediction to a zero tesnor C to zero-padding predcition. I hope C is trainable. Is it allowed?

### Will this change the current api? How?
No
### Who will benefit with this feature?
Some researchers of image processing
Any Other info.

**Comment by alextp** [comment]
I don't see in your code where you use session.run to make the assignments happen. 

Are you using eager execution?

**Comment by chencan15** [comment]
Thanks for your reply!

I'm not using eager execution.

The above code is the neural network i defined, it ruturns C as the prediction of the neural network. 
I name the network as NT, the input of NT is A and the output of NT is C:
C=NT(A)


I assign the prediction to a zero tensor C, the C should be non-zero and trainable. But when i run the following code to print C, all elements in C are still zero.


Here is the code:

with tf.Session() as sess:
       sess.run(tf.global_variables_initializer())
       C_example= sess.run(C, feed_dict={A: train_data_batch})
       C_sum=sess.run(tf.reduce_sum(C_example))
print(C_sum)
print(C)

**Comment by chencan15** [comment]
@alextp

**Comment by alextp** [comment]
Right but when you do c[i].assign without using the return value you create an assignment operation in the graph but never actually execute it.

You have two options to make this code work. The first is in the TF 1.x style adding a lot of control dependencies to build an update op which depends on all assignments; the second is to use [tf.function](https://www.tensorflow.org/alpha/tutorials/eager/tf_function) (which is available in 1.13 but with fewer bugs in nightly) to build your graph and automatically add control dependencies for you.

**Comment by chencan15** [comment]
Sorry for not response promptly. 
Thanks for your reply！

I tried your first options——add a lot of control dependencies. But it still doesn't work. Here is the code:

input_size = A_size * A_size
output_size = B_size * B_size

w = tf.Variable ( tf.random_normal ( [input_size , output_size ] ) )
b = tf.Variable ( 0.0 )



prediction = tf.matmul ( A , w ) + b

C = tf.Variable ( np.zeros ( [ batch_size , C_size * C_size ] ) , True, dtype = tf.float32 )
 
for i in range ( batch_size ) :
         
            temp = tf.Variable ( np.zeros ( [ C_size , C_size ] ) , True , dtype = tf.float32 )
          
            ops_1= tf.assign ( temp [ position_height_index [ i ] * B_size : ( position_height_index [ i ] + 1 ) * B_size, position_width_index [ i ] * B_size : ( position_width_index [ i ] + 1 ) * B_size ] , ( tf.reshape ( prediction [ i ] , [ B_size , B_size ] ) ) )
            
 
           with tf.control_dependencies ( [ ops_1 ] ) :
                      tf.assign ( C [ i ] , ( tf.reshape ( temp, [ 1, C_size * C_size ] ) ) )


Grateful for helping me!

**Comment by alextp** [comment]
there is still nothing depending on the last tf.assign on your code, so it won't execute.

You should switch to tf2 alpha as eager execution fits the style of code you're trying to write much better.
